pydy/tests/test_joints.py

```diff
@@ -0,0 +1,275 @@
+
+        # part 1 assining parent-child relationship
+        self.pinjoint._set_parent_child_rel()
+        assert self.child.parent == self.parent
+        assert self.parent.child == self.child
+
+        # part 2 locating joint point in bodies
+        self.pinjoint._locate_joint_point()
+        child_masscenter = self.child.masscenter
+        parent_masscenter = self.parent.masscenter
+        parent_point = parent_masscenter.locatenew(self.name + '_Point', self.parent_joint_vector)
+        child_point = child_masscenter.locatenew(self.name + '_Point', self.child_joint_vector)
+        assert self.pinjoint.parent_joint_vector == parent_point
+        assert self.pinjoint.child_joint_vector == child_point
+
+        # part 3 aligning parent_axis and child_axis
+        self.join_frames()
+        # sample implementation
+        # self.child.frame.orient(self.parent.frame, 'Axis', [0, self.parent.frame.x])
+        assert cross(self.pinjoint.parent_axis, self.pinjoint.child_axis) == self.child.frame.z
+        assert cross(self.pinjoint.child_axis, self.pinjoint.parent_axis) == - self.parent.frame.z
+        self.align_axes()
+        parent_axis = self.pinjoint.parent_axis
+        child_axis = self.pinjoint.child_axis
+        # sample implementation
+        # perpendicular_axis_in_parent = - cross(parent_axis, child_axis)
+        # angle_between_axes = acos(dot(self.pinjoint.child_axis, self.pinjoint.parent_axis))
+        # self.child.frame.orient(self.parent.frame, 'Axis',
+        #                        [acos(dot(parent_axis, child_axis)/(parent_axis.magnitude() * child_axis.magnitude())),
+        #                         cross(parent_axis, child_axis)])
+        assert acos(dot(parent_axis, child_axis)/(parent_axis.magnitude() * child_axis.magnitude())) == 0
+
+        # part 4 adding angular velocity to child w.r.t parent.
+        assert self.child.frame.ang_vel_in(self.parent.frame) != 0
+
+    def test_pinjoint_apply_joint(self):
+        # apply_joint() should do everything done above by calling specific funtions
+        self.new_pinjoint = PinJoint('pinjoint', self.parent, self.child)
+        self.new_pinjoint._apply_joint()
+        # part 1
+        assert self.child.parent == self.parent
+        assert self.parent.child == self.child
+
+        # part 2
+        child_masscenter = self.child.masscenter
+        parent_masscenter = self.parent.masscenter
+        parent_point = parent_masscenter.locatenew(self.name + '_Point', self.parent_joint_vector)
+        child_point = child_masscenter.locatenew(self.name + '_Point', self.child_joint_vector)
+        assert self.pinjoint.parent_joint_vector == parent_point
+        assert self.pinjoint.child_joint_vector == child_point
+
+        # part 4
+        parent_axis = self.pinjoint.parent_axis
+        child_axis = self.pinjoint.child_axis
+        assert acos(dot(parent_axis, child_axis)/(parent_axis.magnitude() * child_axis.magnitude())) == 0
```

Comment:
If you implement this on vector, you need to change this test to use the method.